test/unit/math/test_ad.hpp

```diff
@@ -1526,6 +1629,93 @@ inline void test_matvar_sum_gradient(MatVar&& A_mv, VarMat&& A_vm,
+ *  Scalar vars.
+ * @tparam VarMat1 A `var_value` with an inner type inheriting from `EigenBase`.
+ * @tparam VarMat2 A `var_value` with an inner type inheriting from `EigenBase`.
+ * @tparam ResultMatVar The resulting type of applying a function to `A_mv1`
+ * and `A_mv2`.
+ * @tparam ResultVarMat The result type of applying a function to `A_vm1` and
+ *  `A_vm2`.
+ * @param A_mv1 A eigen type of vars.
+ * @param A_mv2 A eigen type of vars.
+ * @param A_vm1 A eigen type of vars.
+ * @param A_vm2 A eigen type of vars.
+ * @param A_mv_f The result of a function from applying `A_mv1` and `A_mv2`.
+ * @param A_vm_f The result of a function from applying `A_vm1` and `A_vm2`
+ */
+template <typename ResultVarMat, typename ResultMatVar, typename MatVar1,
+          typename MatVar2, typename VarMat1, typename VarMat2>
+inline void test_matvar_sum_gradient(ResultMatVar& A_mv_f, ResultVarMat& A_vm_f,
+                                     const MatVar1& A_mv1, const MatVar2& A_mv2,
+                                     const VarMat1& A_vm1,
+                                     const VarMat2& A_vm2) {
+  using stan::math::var;
+  var b_mv = sum(A_mv_f);
+  var b_vm = sum(A_vm_f);
+  var final_var = b_mv + b_vm;
+  stan::math::grad(final_var.vi_);
+  if (!stan::math::isfinite(final_var)) {
+    return;
+  }
+  expect_near_rel("var<Matrix> vs Matrix<var> sum intermediate value",
+                  b_vm.val(), b_mv.val(), 1e-12);
+  expect_near_rel("var<Matrix> vs Matrix<var> sum intermediate adjoints",
+                  b_vm.adj(), b_mv.adj(), 1e-12);
+  expect_near_rel_var("lhs var<Matrix> vs Matrix<var> result", A_vm1, A_mv1);
+  expect_near_rel_var("rhs var<Matrix> vs Matrix<var> result", A_vm2, A_mv2);
+  expect_near_rel("var<Matrix> vs Matrix<var> sum result value", A_vm_f.val(),
+                  A_mv_f.val(), 1e-12);
+  expect_near_rel("var<Matrix> vs Matrix<var> sum result adjoints",
+                  A_vm_f.adj(), A_mv_f.adj(), 1e-12);
+}
+
+template <typename ReturnType, typename Type>
+void check_return_type(const ReturnType& ret, const Type& x) {
+  using stan::is_eigen;
+  using stan::is_var_matrix;
+  using stan::math::var_value;
+  using stan::math::test::type_name;
+  if (is_eigen<Type>::value
+      && !(is_eigen<ReturnType>::value
+           || std::is_same<ReturnType, var_value<double>>::value)) {
+    FAIL() << type_name<Type>() << " returns a " << type_name<ReturnType>()
+           << " but should return either an Matrix<var> or a var_value<double>";
+  } else if (is_var_matrix<Type>::value
+             && !(is_var_matrix<ReturnType>::value
+                  || std::is_same<ReturnType, var_value<double>>::value)) {
+    FAIL() << type_name<Type>() << " returns a " << type_name<ReturnType>()
+           << " but should return either an var<Matrix> or a var_value<double>";
+  }
+}
+
+template <typename ReturnType, typename Type1, typename Type2>
+void check_return_type(const ReturnType& ret, const Type1& x, const Type2& y) {
```

Comment:
I think these two functions make sense but can't they be simplified to something like:

`is_same<ReturnType, promote_var_matrix_t<T1, ...>>::value`

Reply:
We could do that but `promote_var_matrix<>` takes an eigen type as the first argument and we would have to deduce that for these somehow. This version is more wordy but EOD I think easier